File: leetcode/algo2/day10/47_permutations_2.cpp
#include <vector>
#include <iostream>
#include <map>

std::vector<std::vector<int>> res{};
std::map<int, int> fre;
std::vector<int> temp;
int target_length;

void dfs()
{
    if (temp.size() == target_length)
    {
        res.push_back(temp);
        return;
    }
    for (auto x : fre)
    {
        if (fre[x.first] > 0)
        {
            temp.push_back(x.first);
            fre[x.first]--;
            dfs();
            fre[x.first]++;
            temp.pop_back();
        }
    }
}
std::vector<std::vector<int>> permuteUnique(std::vector<int> &nums)
{
    target_length = nums.size();
    for (auto x : nums)
        fre[x]++;
    dfs();
    return res;
}

int main()
{
    std::vector<int> nums = {1, 1, 2, 2};
    for (auto x : permuteUnique(nums))
    {
        for (auto y : x)
            std::cout << y << " ";
        std::cout << std::endl;
    }
}

File: leetcode/algo2/day10/47_permutations_2.js
let res = [];
let temp = [];
let fre = new Map();
let target_length;

var dfs = function () {
    if (temp.length === target_length) {
        res.push([...temp]);
        return;
    }
    for (let x of fre.keys()) {
        if (fre.get(x) > 0) {
            temp.push(x);
            fre.set(x, fre.get(x) - 1);
            dfs();
            fre.set(x, fre.get(x) + 1);
            temp.pop();
        }
    }
}

var permuteUnique = function (nums) {
    target_length = nums.length;
    for (const x of nums) {
        if (fre.get(x) === undefined) fre.set(x, 1);
        else fre.set(x, fre.get(x) + 1);
    }
    dfs()
    return res;
};

let nums = [1, 1, 2, 2, 3];
console.log("hii")
for (const x of permuteUnique(nums)) {
    console.log(x);
}